src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,316 @@
+#include <Common/typeid_cast.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Interpreters/ArithmeticOperationsInAgrFuncOptimize.h>
+#include <IO/WriteHelpers.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int UNEXPECTED_AST_STRUCTURE;
+}
+/// scalar values from the first level
+std::pair<ASTs, ASTs> tryToCatchConst(std::string & name, ASTs & argss)
+{
+    ASTs const_num;
+    ASTs not_const;
+
+    for (const auto &ar: argss)
+    {
+        if (const auto * literal = ar->as<ASTLiteral>())
+        {
+            if (literal->value.getType() == Field::Types::Int64 ||
+            literal->value.getType() == Field::Types::UInt64 ||
```

Comment:
One more tabulation for multicolumn if. It's better to make function like `isConstantField(const Field & )` and rewrite this if with `if (isConstantField(literal->value))`. Am I write the condition means that this field is constant? Wethout fucntion name it's not clear.

There're no Decimals here. Why?

Reply:
"There're no Decimals here. Why?" Cause I have no idea how to correctly work with Field in ClickHouse: how you can see, in minOptimize and in maxOptimize I'm always try to get Int128 (auto num = ar->as<ASTLiteral>()->value.get<Int128>();). That's works with Int64, uint64, int128, but I'm doubt, that it would work with decimals.
Would be cool, if you show me how to do it by normal way